File: Python/Medium/3217.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def modifiedList(
        self, nums: List[int], head: Optional[ListNode]
    ) -> Optional[ListNode]:
        # use a set to store illegal numbers
        numbers = set(nums)

        dummy = ListNode(next=head)

        holder = dummy

        while holder.next != None:
            num = holder.next.val
            if num in numbers:
                holder.next = holder.next.next
            else:
                holder = holder.next

        return dummy.next

File: Java/Medium/3217.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode modifiedList(int[] nums, ListNode head) {
        // store the numbers to remove
        Set<Integer> numbers = new HashSet<>();
        for (int i : nums) {
            numbers.add(i);
        }

        // iterate through the ListNodes
        // use a dummy to store the start
        ListNode dummy = new ListNode();
        dummy.next = head;

        int num;
        ListNode holder = dummy;
        while (holder.next != null) {
            num = holder.next.val;
            // if invalid
            if (numbers.contains(num)) {
                holder.next = holder.next.next;
            } else {// need shift my holder to the next one
                holder = holder.next;
            }
        }

        return dummy.next;
    }
}